source/extensions/common/http_wasm/guest.h

```diff
@@ -0,0 +1,231 @@
+
+  // Capabilities which are allowed to be linked to the module. If this is empty, restriction
+  // is not enforced.
+  AllowedCapabilitiesMap allowed_capabilities_;
+
+  std::shared_ptr<GuestHandle> parent_guest_handle_;
+
+  // Used by the uninitialized_guest to enable non-clonable thread local Guest(s) to be constructed.
+  std::string module_bytecode_;
+  std::string module_precompiled_;
+  std::unordered_map<uint32_t, std::string> function_names_;
+
+  std::string vm_configuration_;
+  bool stop_iteration_ = false;
+  FailState failed_ = FailState::Ok; // Guest fatal error
+};
+using GuestSharedPtr = std::shared_ptr<Guest>;
+
+class GuestHandle : public ThreadLocal::ThreadLocalObject,
+                    public std::enable_shared_from_this<GuestHandle> {
+public:
+  explicit GuestHandle(std::shared_ptr<Guest> guest_) : guest_(guest_) {}
+
+  bool canary(const std::shared_ptr<InitializedGuest>& initialized_guest,
+              const GuestHandleCloneFactory& clone_factory);
+
+  void kill() { guest_ = nullptr; }
+
+  std::shared_ptr<Guest>& guest() { return guest_; }
+
+protected:
+  std::shared_ptr<Guest> guest_;
+  std::unordered_map<std::string, bool> initialized_guest_canary_cache_;
+};
+using GuestHandleSharedPtr = std::shared_ptr<GuestHandle>;
+
+std::string makeVmKey(std::string_view vm_id, std::string_view configuration,
+                      std::string_view code);
+
+class InitializedGuestHandle : public std::enable_shared_from_this<InitializedGuestHandle> {
+public:
+  explicit InitializedGuestHandle(std::shared_ptr<GuestHandle> guest_handle,
+                                  std::shared_ptr<InitializedGuest> initialized_guest)
+      : initialized_guest_(initialized_guest), guest_handle_(guest_handle) {}
+  std::shared_ptr<InitializedGuest>& initializedGuest() { return initialized_guest_; }
+  std::shared_ptr<Guest>& guest() { return guest_handle_->guest(); }
+  GuestHandleSharedPtr& wasmHandle() { return guest_handle_; }
+  uint32_t rootContextId();
+
+private:
+  std::shared_ptr<InitializedGuest> initialized_guest_;
+  std::shared_ptr<GuestHandle> guest_handle_;
+};
+
+using InitializedGuestHandleSharedPtr = std::shared_ptr<InitializedGuestHandle>;
+class InitializedGuestHandleSharedPtrThreadLocal : public ThreadLocal::ThreadLocalObject {
+public:
+  InitializedGuestHandleSharedPtrThreadLocal(InitializedGuestHandleSharedPtr handle)
+      : handle_(handle){};
+  InitializedGuestHandleSharedPtr& handle() { return handle_; }
+
+private:
+  InitializedGuestHandleSharedPtr handle_;
+};
+
+using loadGuestCallback = std::function<void(GuestHandleSharedPtr)>;
+bool loadGuest(const InitializedGuestSharedPtr& initialized_guest,
+               const Stats::ScopeSharedPtr& scope, Upstream::ClusterManager& cluster_manager,
+               Event::Dispatcher& dispatcher, Api::Api& api,
+               Envoy::Server::ServerLifecycleNotifier& lifecycle_notifier,
+               loadGuestCallback&& callback);
```

Comment:
I believe this `loadGuestCallback` won't be needed and  we should be able to make `loadGuest` a procedural function (removing `completeCb` etc.

Reply:
`loadGuestCallback` is mainly for setting tls_slot_ which is private member of FilterConfig. Did you mean to set this from `loadGuest()`?
Removed completeCb.